src/pudl/etl.py

```diff
@@ -269,29 +300,46 @@ def etl_epacems(
             "Some timezones may be estimated based on plant state."
         )
 
-    # NOTE: This is a generator for raw dataframes
-    epacems_raw_dfs = pudl.extract.epacems.extract(
-        epacems_settings, Datastore(**ds_kwargs)
-    )
-
-    # NOTE: This is a generator for transformed dataframes
-    epacems_transformed_dfs = pudl.transform.epacems.transform(
-        epacems_raw_dfs=epacems_raw_dfs,
-        pudl_engine=pudl_engine,
-    )
-
     logger.info("Processing EPA CEMS data and writing it to Apache Parquet.")
     if logger.isEnabledFor(logging.INFO):
         start_time = time.monotonic()
 
-    # run the cems generator dfs through the load step
-    for df in epacems_transformed_dfs:
-        pudl.load.df_to_parquet(
-            df,
-            resource_id="hourly_emissions_epacems",
-            root_path=Path(pudl_settings["parquet_dir"]) / "epacems",
-            partition_cols=["year", "state"],
+    if epacems_settings.partition:
+        n_yrs = len(epacems_settings.years)
+        epacems_dir = Path(pudl_settings["parquet_dir"]) / "epacems"
+        with ProcessPoolExecutor() as executor:
+            _ = list(  # Convert results of map() to list to force execution
```

Comment:
I think `map` is the right call here even though it requires this list conversion. I looked a little at `executor.submit()` + `as_completed()` but I think that only speeds things up if there is code executed in this same function after this call to `_etl_one_year_epacems` and if `_etl_one_year_epacems` took different amounts of time to process for different year/state combos. I'm not sure if `as_completed()` is a safety check to ensure the executor has finished all years/states before shutting down? It doesn't make sense to me why the executor would shut down if the tasks aren't done but I've seen `as_completed` included at the end of a function in a lot of places.

If nothing seems to be waiting or hanging during `etl_epacems` then `map` is probably working just fine.

Reply:
There was one time when I was developing this that a process seemed to have stalled out and never came back, and because I hadn't set a `timeout` I eventually had to kill it. Ideally there would be some way to keep retrying in the event that one of the processes hangs or dies. But it only happened once so I didn't focus on it. `map()` seemed like the easiest thing to do since it really is just handing out a bunch of nearly identical work units to the same function. Really I hope we replace this relatively soon with a Dask executor, which will probably have more functionality around dealing with dead subprocesses, since I imagine problems like that are more common in a distributed computing context. Or we'll have Dagster set up with some retry logic.

Reply:
I think the case of the hanging / dead process that you're describing is probably the case for a function like `as_completed`. Agree that if we intend to replace this with a Dask executor soon, then there's no point researching this too much now.

Reply:
If hanging becomes an issue, we can dive into it.